Improve random system generation, load custom systems from JSON

Web-eWorks opened the pull request:
This is a prerequisite PR for the system editor which happens to bring some nice standalone features to the table. I've significantly improved the way custom system definitions are loaded and used in system generation, made several opinionated changes to improve the results of system generation, and added JSON as the preferred format for loading custom systems.

**Custom System Improvements:**

I've mostly focused on reducing the amount of cruft and duplicated effort in custom system handling (as a natural consequence of adding an in-engine editor for custom system files). This results in reproducible round-trip loading and saving of custom system files (in JSON only) and reduces the amount of "knock on" changes tweaking a parameter has on random generation.

As part of this effort, all systems and bodies now receive a seed value for increased determinism in random generation instead of using the system or body index (which would create an implicit dependency on the rest of the sector or system). The number of randomly generated parameters in a system definition are reduced and responsibility for generating "missing" parameters in a Lua system definition file is now handled during the loading phase rather than the system generation phase. (Json files require that all parameters are specified and does not perform standalone generation.)

**System Generation Changes:**

SystemBodies are now responsible for calculating atmospheric parameters, moved from Planet.cpp. Additionally, SystemBody now has a proper parameterization for Argument of Periapsis, allowing it to fully express a proper orbit.

Bodies now generate in a more realistic fashion in an attempt to relax the behavior of cramming 5-10 gas giants within 5 AU of the primary star. Several scaling factors are applied to body generation, more satellites are generated around gas giants (referencing Jupiter), and a factor is applied to prevent extremely dense stars from generating large numbers of high-mass satellites in a very small orbit.

**JSON Custom Systems:**

As part of the system editor work, I've added a JSON-based format for loading and saving custom system definitions. This enables repeatable round-trip serialization, potentially higher performance with 100s of custom systems, and is a much easier and less bug-prone format for the engine to import and export. This comes at the cost of the ability to hand-edit the file (unless you want to do fixed-point format conversion by hand as well), though the files remain readable and easy to diff.

This new format makes Lua-based custom systems **officially deprecated**, as it is much better at representing all of the parameters that make up a system. (When compared to the Lua format, where 50% or more of body parameters were completely implicit and changed based on the order the system was loaded in.)

This new format is 100% **fully editable** with the upcoming System Editor, which provides a fully visual edit interface to adjust all parameters of a system, provides procedurally-assisted body generation, and can calculate derived parameters like surface gravity and atmosphere height in real-time. Plus undo. Who doesn't like undo?

Sneak peek:

https://github.com/pioneerspacesim/pioneer/assets/4218491/8d0a9a54-00ec-4aa9-8188-c046f6cbe6ff

(MSAA has been added since...)

The JSON-based format is not end-user testable in this PR without the system editor - this code has been developed in the system editor branch however, and is reasonably bug-free.

### TODO:
- [x] Autoload custom systems from JSON during galaxy generation
- [x] Load "partial" custom systems from JSON (for e.g. local_stars, hyg database)
- [x] Override "partial" custom system definitions with fully-defined custom systems (e.g. local_stars.lua)
- [x] Prefer JSON custom systems to Lua custom systems

I'm requesting "review" from @fluffyfreak and @Gliese852 to take a quick look at the description and PR and let me know feedback/thoughts on the general direction. Overall, I'm pretty happy with the code quality and effect of the PR, but more eyes are always welcome!

fluffyfreak commented:
I'll get on it, though it might take me a week 👍

impaktor commented:
Was going to test it, but crashes (window frame is rendered, but then crashes):

```
Info: Lua 5.2
Info: Found mod Vgr_battlecruiser, enabled: true
Info: 	source: /home/karlf/.pioneer/mods/Vgr_battlecruiser.zip
Info: Found mod longnose0.4, enabled: true
Info: 	source: /home/karlf/.pioneer/mods/longnose0.4.zip
Info: Loaded mods:
Info: 	Vgr_battlecruiser
Info: 	longnose0.4
Warning: couldn't read language file 'lang/core/.json'
Warning: couldn't load language resource core/, trying core/en
Info: --------------------
Info: SDL Version (build) 2.28.2
Info: SDL Version (dynamic) 2.28.2
Info: SDL Versions match
Info: SDL_image Version (build): 2.6.3
Info: SDL_image Version (dynamic): 2.6.3
Info: SDL_image Versions match
Info: Assimp Version: 5.2.0
Info: FreeType Version: 2.13.1
Info: GLEW dynamic version: 2.0.0
Info: --------------------
Info: 
Info: Initialized OpenGL 3.1, with extensions, renderer
Info: Created shader vtxColor (address=0x560b7346cbd0)
Info: Initializing joystick subsystem.
Info: Loaded PiGui fonts from disk:
Info: font console:0
Info:   - Inpionata.ttf 1.0
Info:   - DejaVuSansMono.ttf 1.2
Info:   - wqy-microhei.ttc 1.0
Info:   - icons/icons.svg 16x19
Info: font icons:0
Info:   - PionilliumText22L-Medium.ttf 1.0
Info:   - icons/icons.svg 16x19
Info: font orbiteer:0
Info:   - Orbiteer-Bold.ttf 1.0
Info:   - DejaVuSans.ttf 1.2
Info:   - wqy-microhei.ttc 1.0
Info:   - icons/icons.svg 16x19
Info: font pionillium:0
Info:   - PionilliumText22L-Medium.ttf 1.0
Info:   - DejaVuSans.ttf 0.9285714
Info:   - wqy-microhei.ttc 1.0
Info:   - icons/icons.svg 16x19
Info: Created shader ui (address=0x560b7346e4b0)
Warning: No SVG data available to rasterize icon font icons/icons.svg
Warning: No SVG data available to rasterize icon font icons/icons.svg
```

Web-eWorks commented:
A backtrace would be useful :wink:

Unsure about the crashing - the code should be functional but was cherry-picked out of a larger branch so it's possible I missed an implicit dependency that's present in the full system-editor branch.

fluffyfreak commented on src/galaxy/CustomSystem.cpp:
```diff
@@ -314,6 +318,49 @@ static luaL_Reg LuaCustomSystemBody_meta[] = {
 	{ 0, 0 }
 };
 
+void CustomSystemBody::LoadFromJson(const Json &obj)
```
Are these so different from SystemBody that we can't have a common base class implement this?

Web-eWorks replied:
I was trying to do that when I first wrote this code - at that time the combination of loading the system hierarchy from Lua and SystemBody being unable to be copy-constructed made it infeasible, but I should now be able to implement a common copyable ancestor holding the variables.

Web-eWorks replied:
Implemented a common "data container" used by both CustomSystem and SystemBody.

fluffyfreak reviewed (approved):
I think this all looks good, I'm guessing savegame bump and that it cahnges the universe quite a lot.

Approved

fluffyfreak commented:
I get a crash in testing x64|Debug looks like it's trying to load CityOnPlanet `Dome1` 

> ` 	vcruntime140d.dll!00007ffd0ff11416()	Unknown
 	pioneer.exe!std::_Char_traits<char,int>::copy(char * const _First1, const char * const _First2, const unsigned __int64 _Count) Line 67	C++
 	pioneer.exe!std::string::_Construct_lv_contents(const std::string & _Right) Line 2917	C++
 	pioneer.exe!std::string::basic_string<char,std::char_traits<char>,std::allocator<char>>(const std::string & _Right) Line 2421	C++
 	pioneer.exe!std::_Default_allocator_traits<std::allocator<std::string>>::construct<std::string,std::string const &>(std::allocator<std::string> & __formal, std::string * const _Ptr, const std::string & <_Args_0>) Line 683	C++
 	pioneer.exe!std::vector<std::string,std::allocator<std::string>>::_Emplace_reallocate<std::string const &>(std::string * const _Whereptr, const std::string & <_Val_0>) Line 777	C++
 	pioneer.exe!std::vector<std::string,std::allocator<std::string>>::emplace_back<std::string const &>(const std::string & <_Val_0>) Line 734	C++
 	pioneer.exe!std::vector<std::string,std::allocator<std::string>>::push_back(const std::string & _Val) Line 744	C++
 	pioneer.exe!Assimp::IOSystem::PushDirectory(const std::string & path) Line 286	C++
 	[External Code]	
 	pioneer.exe!Assimp::Importer::ReadFile(const std::string & pFile, unsigned int pFlags) Line 645	C++
 	pioneer.exe!SceneGraph::Loader::LoadMesh(const std::string & filename, const std::vector<SceneGraph::AnimDefinition,std::allocator<SceneGraph::AnimDefinition>> & animDefs) Line 353	C++
 	pioneer.exe!SceneGraph::Loader::CreateModel(SceneGraph::ModelDefinition & def) Line 260	C++
 	pioneer.exe!SceneGraph::Loader::LoadModel(const std::string & shortname, const std::string & basepath) Line 203	C++
 	pioneer.exe!SceneGraph::Loader::LoadModel(const std::string & filename) Line 136	C++
>	pioneer.exe!ModelCache::FindModel(const std::string & name) Line 25	C++
 	pioneer.exe!Pi::FindModel(const std::string & name, bool allowPlaceholder) Line 1172	C++
 	pioneer.exe!CityOnPlanet::LoadBuildingType(std::basic_string_view<char,std::char_traits<char>> key, const nlohmann::basic_json<std::map,std::vector,std::string,bool,__int64,unsigned __int64,double,std::allocator,nlohmann::adl_serializer> & buildingDef, CityOnPlanet::BuildingType & out) Line 106	C++
 	pioneer.exe!CityOnPlanet::LoadCityFlavour(const FileSystem::FileInfo & file) Line 214	C++
 	pioneer.exe!CityOnPlanet::Init() Line 245	C++
 	[External Code]	
 	pioneer.exe!StartupScreen::RunNewLoader() Line 628	C++
 	pioneer.exe!StartupScreen::Update(float deltaTime) Line 607	C++
 	pioneer.exe!Application::Run() Line 195	C++
 	pioneer.exe!SDL_main(int argc, char * * argv) Line 180	C++
 	[External Code]	
`

fluffyfreak commented:
For me, on Windows x64|Debug in VS2019 using CMake it fails with Assimp loading `models/buildings/vlastan/Dome1LQ.obj`

Web-eWorks commented:
@fluffyfreak ...I haven't rebased this PR ontop of the one that fixes Assimp in debug mode. That'd be the cause of the crash.

Web-eWorks commented:
Rebased onto master (should now run in Win64 Debug), and refactored CustomSystemBody to use a new shared base class `SystemBodyData`. This massively simplifies code in both CustomSystem and StarSystemGenerator - thanks for the reminder/nudge in that direction!

Still procrastinating on hooking up JSON custom systems to galaxy generation, but will try to get that done soonish.

fluffyfreak commented:
Yep that all works for me in Debug and Release 👍

impaktor commented:
> A backtrace would be useful

I just tried again, on current state [47f484b](https://github.com/pioneerspacesim/pioneer/pull/5622/commits/47f484bdd810027060c4094db2ef7087b466ab61). Get no trace.
```
gdb ./build/editor                                                                                  
GNU gdb (GDB) 13.2
(gdb) bt
No stack.
(gdb) quit
```

Web-eWorks commented:
@impaktor you're trying to run the editor, which a) isn't functional in this PR (only the system generation changes are available for testing here), and b) requires you to pass `--system` to enter the System Editor mode.

Web-eWorks commented:
Implemented all TODO items and migrated `00_sol.lua` to JSON for testing purposes. Feel free to change names etc. in the file and ensure it's being loaded properly when starting a game at Mars.

If no further issues arise on this PR, will most likely merge it by end-of-week.

Gliese852 commented on src/galaxy/CustomSystem.cpp:
```diff
@@ -660,14 +752,189 @@ void CustomSystemsDatabase::Load()
+	PROFILE_SCOPED()
+
+	assert(!s_activeCustomSystemsDatabase);
+	s_activeCustomSystemsDatabase = this;
```
Maybe we could explicitly pass this object to Lua? Otherwise it looks a little suspicious.

Web-eWorks replied:
Were we intending to continue supporting lua system definitions long-term I'd consider it; however I don't intend to sink any additional time into supporting lua custom system definitions and consider all lua-related code deprecated.

Gliese852 reviewed (commented):
Thank you for inviting me to look at your code, it was very interesting and informative, and I managed to find a few things to pick on!

Web-eWorks commented on data/systems/custom/00_sol.json:
```diff
@@ -0,0 +1,2762 @@
+			"life": "f0/0",
+			"mass": "f0/0",
+			"metallicity": "f0/0",
+			"name": "Cydonia",
```
I'd much rather some bodies have well-defined but completely unused fields, than have them be underdefined as our current bodies are. Consider it future-proofing :smile:

Web-eWorks commented on src/galaxy/CustomSystem.cpp:
```diff
@@ -571,10 +587,12 @@ static int l_csys_add_to_sector(lua_State *L)
 	(*csptr)->sectorY = y;
 	(*csptr)->sectorZ = z;
-	(*csptr)->pos = vector3f(*v);
+	(*csptr)->pos = vector3f(*v) * Sector::SIZE; // NOTE: lua uses 0..1 interval inside a sector cell
```
The raw value used by the entire codebase for system positions within a sector is in lightyears in the range [0..8) ly as sector sizes are defined. It's much easier and more intuitive for the user to edit the sector position in physical units, rather than an abstract scalar of 0..1 which requires an *additional* hand-calculation by the user to place the system in the sector. This multiplication was moved here only to maintain backwards compatibility with existing lua files.

Gliese852 replied:
On the other hand, you need to know exactly the sector size to understand where the system is located. _Personally_, coordinate 0..1 seems to me more intuitive, in addition, it is consistent with the coordinates of the sector - where the integer just means 1 sector - 8 light years. But this doesn’t make any fundamental difference to me.

Web-eWorks commented on data/systems/partial/02_local_stars.json:
```diff
@@ -0,0 +1,3 @@
+[
```
The only thing I can think of that might be better would be `data/systems/stars/02_local_stars.json`, but that doesn't communicate well the partial/"stub" nature of the definitions.

Gliese852 replied:
I guess we just need to get used to this name, and maybe describe it somewhere, most likely people who don’t read code will encounter this concept.

Web-eWorks commented on src/JsonUtils.cpp:
```diff
@@ -538,22 +556,22 @@ void from_json(const Json &obj, fixed &f)
 			throw Json::type_error::create(320, "cannot pickle string to fixed point number");
 
-		f = fixed(numerator, denominator);
+		f = fixed(numerator << f.FRAC | denominator);
```
`integer` and `fractional` are perfectly fine, I'll see about editing the variable names to make it clear. I was only concerned with getting fixed-point round-trip working at the time I made these changes.

Web-eWorks commented on src/Random.h:
```diff
@@ -233,6 +246,25 @@ class Random : public RefCounted {
+
+	// interval (mean - stddev, mean + stddev)
+	// this is not a true gaussian distribution
+	inline fixed NormFixed(fixed mean, fixed stddev)
```
`stddev` was only chosen because the function was attempting to mimic the API of `Normal(mean, stddev)`. I don't mind re-naming it, as the parameter is indeed intended to be a hard limit on the maximum value produced by the function.

That graph for `FixedFixedFixed/3` looks significantly better than the approximation I came up with, thank you! I kludged together the existing SFixed() approximation with the help of an online probability visualization tool that wasn't very good.

Web-eWorks replied:
Out of curiosity, what is the graph of two calls to the new implementation multiplied against each other compared to the graph of SFixed(4) or SFixed(5)? I'm entirely replacing SFixed() with the new NormFixed() implementation, and I'd like to know how much error compared to the previous iteration I'm introducing here.

Gliese852 replied:
![изображение](https://github.com/pioneerspacesim/pioneer/assets/18342621/382449f2-2b44-418a-8779-a16723f6b48d)

I put all three graphs (scaled down 10 times vertically), you can see that `NormFixed()*NormFixed()` is the sharpest, although the peak is the same as that of `SFixed(5)`.
If maybe you want to play around with it, and you're not afraid of your eyes bleeding, I pushed my [test mess](https://github.com/Gliese852/gles_imgui/tree/histogram), the shape of the graphs is defined [here](https://github.com/Gliese852/gles_imgui/blob/571750eb739760bbec710b6ed21e086301af8ae8/main.cpp#L193).
(In order for the graph to grow, you need to move the mouse, since redrawing occurs only in the event handler)

impaktor replied:
Perhaps useful to subtract the two, and only show the (relative?) differences? (Depending on what is interesting; I'm not really following the details of the discussion)

Web-eWorks replied:
Thanks! That matched what I was able to determine using JupyterLab; I wound up with a slightly different term to control body inclination, though I may still use `NormFixed() * NormFixed()`.

Web-eWorks commented on src/galaxy/StarSystemGenerator.cpp:
```diff
@@ -357,220 +357,144 @@ void StarSystemLegacyGeneratorBase::PickRings(SystemBody *sbody, bool forceRings
 
 void StarSystemCustomGenerator::CustomGetKidsOf(RefCountedPtr<StarSystem::GeneratorAPI> system, SystemBody *parent,
-	const std::vector<CustomSystemBody *> &children, int *outHumanInfestedness, Random &rand)
+	const std::vector<CustomSystemBody *> &children, int *outHumanInfestedness)
```
I'm going to defer this feedback for another, later pass on system generation, as it should now be possible to load all bodies in a single pass and only do hierarchical descent to set parent/children pointers on the loaded bodies. That would obviate the need for the recursive `CustomGetKidsOf` function entirely.

Web-eWorks commented on src/galaxy/StarSystemGenerator.cpp:
```diff
@@ -357,220 +357,144 @@ void StarSystemLegacyGeneratorBase::PickRings(SystemBody *sbody, bool forceRings
-			kid->m_orbit.SetShapeAroundPrimary(csbody->semiMajorAxis.ToDouble() * AU, parent->GetMass(), csbody->eccentricity.ToDouble());
+		// parent gravpoint mass = sum of masses of its children
+		if (parent->GetType() == SystemBody::TYPE_GRAVPOINT) {
+			if (kid->m_type >= SystemBody::TYPE_STAR_MIN && kid->m_type <= SystemBody::TYPE_STAR_MAX)
```
Existing code copy/pasted... yes, that'd be a much more concise way of handling it.

Web-eWorks commented on src/galaxy/StarSystemGenerator.cpp:
```diff
@@ -880,8 +843,20 @@ static fixed mass_from_disk_area(fixed a, fixed b, fixed max)
+	return (b * b * one_max_b) - (a * a * one_max_a);
 }
 
 static fixed get_disc_density(SystemBody *primary, fixed discMin, fixed discMax, fixed percentOfPrimaryMass)
```
The case of discMax <= discMin should be handled at a higher level in all cases. I believe I added in appropriate guards; I've certainly not triggered the `assert(b >= a)` case slightly higher in the same function in the current iteration of the code.

Gliese852 replied:
`assert(b >= a)` passes when `b == a`.
you can see that
`discMax = std::max(discMax, discMin);`
 just makes them equal

Web-eWorks replied:
I've partially addressed this in `get_disc_density`, and going to claim "it was like that before I got there" for the remaining semantic problems (as I don't have a good solution for it at this time). If you're better able to make heads or tails of the function and its context to further prevent divide-by-zero I'd welcome that as a separate PR.

Web-eWorks commented on src/galaxy/StarSystemGenerator.cpp:
```diff
@@ -1046,6 +1033,114 @@ void StarSystemRandomGenerator::MakePlanetsAround(RefCountedPtr<StarSystem::Gene
+		fflush(Log::GetLog()->GetLogFileHandle());
+	}
+
+	assert(mass >= 0);
```
Existing code. The assert could probably be removed, as I think I've managed to remove or reduce to insignificance the incidence rate of mass overflow with the current iteration of system generation.

Gliese852 replied:
I meant what you commented just above
`// mass = fixed(Sint64(0x7fFFffFFffFFffFFull));`
in the existing code the mass became maximum fixed if it turned out to be negative.

Web-eWorks commented on src/galaxy/StarSystemGenerator.cpp:
```diff
@@ -1683,17 +1779,15 @@ void PopulateStarSystemGenerator::SetSysPolit(RefCountedPtr<Galaxy> galaxy, RefC
 	}
 
-	if (customSystem && !customSystem->want_rand_lawlessness)
-		sysPolit.lawlessness = customSystem->lawlessness;
```
Should be set correctly in `StarSystemCustomGenerator` (at [line 453](https://github.com/pioneerspacesim/pioneer/blob/56bdbe95ad4daed90de7a9276b33e9bc9d9053a9/src/galaxy/StarSystemGenerator.cpp#L453)). If the lawlessness from custom systems is being overwritten here, that's a problem that should be corrected.

Gliese852 replied:
The thing is that `PopulateStarSystemGenerator` is applied [last](https://github.com/pioneerspacesim/pioneer/blob/1957b8d8e586ae5ec44a495406168645e8c2b146/src/galaxy/GalaxyGenerator.cpp#L59) and, if I’m not mistaken, it overwrites the value.

Web-eWorks commented on src/galaxy/SystemBody.cpp:
```diff
@@ -6,33 +6,238 @@
+		(surfaceGravity_g * gasMolarMass / (GAS_CONSTANT_R * lapseRate_L))); // in ATM since p0 was in ATM
+}
+
+double SystemBody::GetAtmDensity(double altitude, double pressure) const
```
`GetAtmSurfaceDensity` returns the sea-level atmospheric density given the average temperature of the body. `GetAtmDensity` returns the density of the atmosphere at the given altitude and pressure level, in kg/m^3. At `altitude = 0, pressure = surfacePressure`, this function should return the same as `GetAtmSurfaceDensity`.

Gliese852 replied:
Indeed, now I see it, thanks!

Web-eWorks commented on src/galaxy/SystemBody.cpp:
```diff
@@ -6,33 +6,238 @@
+{
+	// temperature at height
+	const double lapseRate_L = CalcSurfaceGravity() / GetSpecificHeat(GetSuperType()); // deg/m
+	return double(GetAverageTemp()) - lapseRate_L * altitude;
```
If the minuend and subtrahend coincide, the body either has a surface temperature of zero degrees Kelvin or the queried altitude is outside of the atmosphere, at which point the atmospheric temperature is defined as zero. The divide-by-zero should not be an issue; any case where this returns 0 should be caught by a higher-level function checking the input parameter to see if it is within the body's atmosphere or not.

Gliese852 replied:
Indeed, now I see, `m_atmosRadius` is calculated in such a way as to protect from this!

Web-eWorks commented:
@Gliese852 thank you very much for the detailed and comprehensive review of the code! I'll make the appropriate changes to address the review feedback soon - I've left explanations/comments where needed; feel free to ask further questions if I've not answered something clearly / satisfactorily.

Web-eWorks commented on src/galaxy/StarSystemGenerator.cpp:
```diff
@@ -357,220 +357,144 @@ void StarSystemLegacyGeneratorBase::PickRings(SystemBody *sbody, bool forceRings
+	system->SetSysPolit(sysPolit);
+
+	if (customSys->IsRandom())
+		return false;
```
Yes, this return value controls whether the system is treated as "fully custom" and no further random generation is run (sets `config->isCustomOnly`).

Web-eWorks commented on src/galaxy/StarSystemGenerator.cpp:
```diff
@@ -1046,6 +1033,114 @@ void StarSystemRandomGenerator::MakePlanetsAround(RefCountedPtr<StarSystem::Gene
+		// Increment the initial periapsis range by a value that falls off the
+		// further towards the disc edge we are if orbiting a star
+		if (primary->GetSuperType() == SystemBody::SUPERTYPE_STAR) {
+			fixed bump_factor = (fixed(1, 1) - min_slice / max_slice);
```
Good catch, that was supposed to be `discMax`.

Web-eWorks commented:
I've addressed all review feedback comments except regarding the `assert(b >= a)` issue. I'm unsure at this time how to address that problem; I've added a minor patch to `get_disc_density` that prevents a divide by zero but in general I'm unsure what the original intended purpose or application of the function was supposed to be.

Web-eWorks commented:
I've also added a few extra syllable fragments to the name generator lists in a very non-exhaustive fashion, which hopefully improve the uniqueness of generated system names slightly. (It also means we can generate systems named Dune now, which was impossible before.)
